compiler/src/service/compile.rs

```diff
@@ -0,0 +1,564 @@
+
+                task(self);
+
+                // Will never trigger compilation
+                false
+            }
+            // Handle memory events.
+            CompilerInterrupt::Memory(event) => {
+                log::debug!("CompileActor: memory event incoming");
+
+                // Emulate memory changes.
+                let mut files = HashSet::new();
+                if matches!(event, MemoryEvent::Sync(..)) {
+                    files = self.estimated_shadow_files.clone();
+                    self.estimated_shadow_files.clear();
+                }
+                match &event {
+                    MemoryEvent::Sync(event) | MemoryEvent::Update(event) => {
+                        for path in event.removes.iter().map(Deref::deref) {
+                            self.estimated_shadow_files.remove(path);
+                            files.insert(path.into());
+                        }
+                        for path in event.inserts.iter().map(|e| e.0.deref()) {
+                            self.estimated_shadow_files.insert(path.into());
+                            files.remove(path);
+                        }
+                    }
+                }
+
+                // If there is no invalidation happening, apply memory changes directly.
+                if files.is_empty() && self.dirty_shadow_logical_tick == 0 {
+                    self.apply_memory_changes(event);
+
+                    // Will trigger compilation
+                    return true;
+                }
+
+                // Otherwise, send upstream update event.
+                // Also, record the logical tick when shadow is dirty.
+                self.dirty_shadow_logical_tick = self.logical_tick;
+                send(Notify(NotifyMessage::UpstreamUpdate(
+                    crate::vfs::notify::UpstreamUpdateEvent {
+                        invalidates: files.into_iter().collect(),
+                        opaque: Box::new(TaggedMemoryEvent {
+                            logical_tick: self.logical_tick,
+                            event,
+                        }),
+                    },
+                )));
+
+                // Delayed trigger compilation
+                false
+            }
+            // Handle file system events.
+            CompilerInterrupt::Fs(event) => {
+                log::debug!("CompileActor: fs event incoming {:?}", event);
+
+                // Handle file system event if any.
+                if let Some(mut event) = event {
+                    // Handle delayed upstream update event before applying file system changes
+                    if let FilesystemEvent::UpstreamUpdate { upstream_event, .. } = &mut event {
+                        let event = upstream_event.take().unwrap().opaque;
+                        let TaggedMemoryEvent {
+                            logical_tick,
+                            event,
+                        } = *event.downcast().unwrap();
```

Comment:
downcast().unwrap()? seems that u r pretty sure what the underlying type is. It is possible to make it static? box<dyn any> is super ugly